### DISCUSSION TITLE
Performance with large forms

### DISCUSSION THREAD

**Discussion by islami00** [react-hook-form/react-hook-form#9446] [Q&A]
Hello,

I really like this form library however I've been running into issues when working with large forms. Specifically, when resetting or updating a lot of fields.

My use case involves:

1.  a group input that edits other inputs. This becomes especially hairy with more than 2000 inputs as it lags for more than a second. Do see the following code sandbox for an example (note that in my case I use `setValue` with `shouldDirty` set as true but the end result is the same as the input is really delayed):
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-scmrwi?file=/src/index.tsx

  Note that the sandbox has 3 examples:
    - One with `useFieldArray` and `update` (the ideal case. `allAtOnce.tsx`), 
    - without `useFieldArray`(assuming that is the issue) and `setValue` and only one field rendered,  `WithSetValue.tsx`
    - and another with `useFieldArray` using `setValue` to update with only 20 inputs rendered.  `WithAFew.tsx`

2. Another use case I have is keeping the dirty state when the form is being saved. I have gone with the approach of resetting optimistically when I submit, then rolling back when an error happens. In both cases, the reset of the form is expensive, taking 2seconds.

I have tried:

1. Rendering less items.(The performance is pretty much the same, I would assume this is independent of the inputs registered because I am calling update on all regardless)

I'm also wondering if a possible `updateAll` function could be given for this use case for `useFieldArray` or batching of checks for isdirty?  This would allow bulk updating with minimal cost.

> I understand that this is may not be unclear as it's kind of a dump of the issues I've been facing for a week now, please let me know if you need any further clarification.

**Comment by FPDK** [comment]
Hi,

I've written two possible solutions [here](https://codesandbox.io/s/react-hook-form-usefieldarray-forked-zi3woj?file=/src/allAtOnce.tsx):
`onChangeCallback1` will update the fields as you type. If you type fast there will still be some lag.
`onChangeCallback2` waits until you've finished typing, before updating the fields.

The reason I'm wrapping the updates in setTimeout should be explained in [this article](https://medium.com/@Rahulx1/understanding-event-loop-call-stack-event-job-queue-in-javascript-63dcd2c71ecd). 

I hope this helps :)

**Reply by islami00** [reply]
Hi,
Thanks for providing these solutions.
For my use case, `onChangeCallback1` would be the closest (there is a preexisting implementation with `setState` and it updates immediately).

Although `onChangeCallback1` solves the problem of updating and the form is faster, it however provides a worse experience viewing the update on the other inputs relative to the `setState` implementation as the user would expect the updates to come as they type.

I've reflected on this and I'll have use setState for the list until support for this is better, I will however keep the discussion open to provide a baseline for this improvement

**Comment by islami00** [comment]
To update this, I tried implementing the setState version of my form with `get` and `set` (because I love the typesafety this library guarantees), however, I still came across the same performance issue. 
Upon further investigation, the bottleneck in the `setState` call was set. As I recall, `set` is implemented recursively, and that would be costly if large arguments are passed to it (which is what I was doing) as a large amount of memory would be required for each recursive call to `set`.

Not sure how accurate this is, but I wanted to share regardless
